statefun-sdk-go/v3/pkg/statefun/context.go

```diff
@@ -76,6 +81,12 @@ func (s *statefunContext) Storage() AddressScopedStorage {
 	return s.storage
 }
 
+func (s *statefunContext) WithContext(ctx context.Context) Context {
+	newContext := *s
+	newContext.Context = ctx
+	return &newContext
```

Comment:
Will this then share the mutex across both contexts? Reading [this thread](https://groups.google.com/g/golang-nuts/c/imxjBLNJ9OY), it seems like another approach would be to lock the mutex, copy the protected fields into a new object (which holds a new mutex), then unlock. We may also want to copy the other pointer fields like caller, storage, address. wdyt?

Reply:
Unless I'm mistaken, the initial assignment operation:

```
newContext := *s
```

... will shallow copy all the fields from the existing `statefunContext` to the new one, including the caller (a pointer), self (not a pointer), storage (a pointer), address (a pointer), response (a pointer), and the mutex (a pointer). 

This means that both contexts will share the same mutex and response, which is what we want, I think? The mutex is used to serialize access to fields in the response, i.e.:

```
s.Lock()
s.response.OutgoingMessages = append(s.response.OutgoingMessages, invocation)
s.Unlock()
```

My read of the linked thread is that it describes how to copy a mutex when it is directly embedded (i.e. not a pointer); in this case, we have the option to make it a pointer and sidestep the problem. But please let me know if you disagree. Thanks.

Reply:
Ah yeah, I was thinking we want this to be a full, independent copy of the Context, without shared state (like in `http.Request`). I don't have enough "context" here to know if that's a requirement / what possible situations would arise if the state is shared.

Reply:
@sjwiesman ?

Reply:
Ah, I understand what you're saying.

I think that, with regard to the mutex and the response, we need the old and new contexts to share the same state, because both contexts need to update the same response object, i.e. the one (and only) response object that will be returned from the invocation of the stateful function.

I think that also holds true for the storage; there is only one storage instance that applies for the entire stateful-function call, so both contexts need to reference it. (And the storage operations are protected with a sync.RWMutex, so this should be safe.)

So that just leaves the caller. My understanding is that this is a pointer in order that it can be nil, to handle the case where there is no caller, but that it is also intended to be read-only. Even without the addition of `WithContext`, two consumers of the same statefun.Context could conceivably step on each others' toes by modifying the `Address` returned from `Caller`. I think the assumption is that the returned `Address` will be read-only, and with that assumption, we wouldn't be making anything worse by shallow copying vs. deep copying.

But, deep copying the caller `Address` would certainly be possible.

@sjwiesman @tillrohrmann Do you have any thoughts on this?

Reply:
The statefun context very explicitly has shared state. It is effectively building the return payload from the batch of messages. Self is static for all messages in the batch. Caller is message specific, but from user code perspective both are immutable for the full invocation.